File: src/Exporter.NX.Translate/SketchTranslator.cs
// SPDX-License-Identifier: GPL-2.0-only
using System;
using System.Collections.Generic;
using Oblikovati.Exporter.NX.Model;
using Oblikovati.Exporter.NX.Recipe;

namespace Oblikovati.Exporter.NX.Translate
{
    /// <summary>
    /// Translates one NX sketch into an Oblikovati <see cref="SketchData"/>: shared
    /// points (via <see cref="SharedPointBuilder"/>), curve entities, geometric
    /// constraints, and parameter-linked dimensions. Lengths convert from the IR's
    /// millimetres to the recipe's centimetre database unit.
    /// </summary>
    public sealed class SketchTranslator
    {
        private const double MmToCm = 0.1;

        private readonly IdAllocator _ids;
        private readonly ExportReport _report;

        public SketchTranslator(IdAllocator ids, ExportReport report)
        {
            _ids = ids ?? throw new ArgumentNullException(nameof(ids));
            _report = report ?? throw new ArgumentNullException(nameof(report));
        }

        public SketchData Translate(NxSketch sketch, int sketchId)
        {
            var points = new SketchPointTable();
            points.Build(sketch, _ids);

            var data = new SketchData
            {
                Id = sketchId,
                Name = sketch.Name.Length == 0 ? null : sketch.Name,
                Plane = TranslatePlane(sketch),
            };
            foreach (PointData p in points.Points)
            {
                data.Points.Add(p);
            }

            var entityIds = AddEntities(sketch, points, data);
            AddConstraints(sketch, points, entityIds, data);
            AddDimensions(sketch, points, entityIds, data);
            return data;
        }

        private static PlaneData TranslatePlane(NxSketch sketch)
        {
            return new PlaneData
            {
                Origin = Scale(sketch.Origin),
                XAxis = (double[])sketch.XAxis.Clone(),
                YAxis = (double[])sketch.YAxis.Clone(),
            };
        }

        // Allocates an entity id per curve (after points) and emits its EntityData.
        private Dictionary<long, int> AddEntities(NxSketch sketch, SketchPointTable points, SketchData data)
        {
            var entityIds = new Dictionary<long, int>();
            foreach (NxCurve curve in sketch.Curves)
            {
                int id = _ids.Next();
                entityIds[curve.Id] = id;
                data.Entities.Add(BuildEntity(id, curve, points));
            }

            return entityIds;
        }

        private static EntityData BuildEntity(int id, NxCurve curve, SketchPointTable points)
        {
            var entity = new EntityData
            {
                Id = id,
                Kind = KindName(curve.Kind),
                Construction = curve.Construction ? true : (bool?)null,
            };
            switch (curve.Kind)
            {
                case NxCurveKind.Line:
                    entity.Points.Add(points.PointId(new NxPointRef(curve.Id, NxCurvePointRole.Start)));
                    entity.Points.Add(points.PointId(new NxPointRef(curve.Id, NxCurvePointRole.End)));
                    entity.Centerline = curve.Centerline ? true : (bool?)null;
                    break;
                case NxCurveKind.Circle:
                    entity.Points.Add(points.PointId(new NxPointRef(curve.Id, NxCurvePointRole.Center)));
                    entity.Radius = curve.Radius * MmToCm;
                    break;
                default: // Arc
                    entity.Points.Add(points.PointId(new NxPointRef(curve.Id, NxCurvePointRole.Center)));
                    entity.Points.Add(points.PointId(new NxPointRef(curve.Id, NxCurvePointRole.Start)));
                    entity.Points.Add(points.PointId(new NxPointRef(curve.Id, NxCurvePointRole.End)));
                    entity.Ccw = curve.Ccw ? true : (bool?)null;
                    break;
            }

            return entity;
        }

        private void AddConstraints(
            NxSketch sketch, SketchPointTable points, Dictionary<long, int> entityIds, SketchData data)
        {
            foreach (NxSketchConstraint c in sketch.Constraints)
            {
                ConstraintData? row = BuildConstraint(c, points, entityIds);
                if (row != null)
                {
                    data.Constraints.Add(row);
                }
            }
        }

        private ConstraintData? BuildConstraint(
            NxSketchConstraint c, SketchPointTable points, Dictionary<long, int> entityIds)
        {
            var row = new ConstraintData { Kind = ConstraintName(c.Kind) };
            switch (c.Kind)
            {
                case NxConstraintKind.Coincident:
                    // Distinct endpoints joined by an explicit coincidence (engine format).
                    row.Points.Add(points.PointId(c.Points[0]));
                    row.Points.Add(points.PointId(c.Points[1]));
                    return row;
                case NxConstraintKind.Horizontal:
                case NxConstraintKind.Vertical:
                    // NX applies these to a line; Oblikovati constrains the line's two endpoints.
                    long line = c.Curves[0];
                    row.Points.Add(points.PointId(new NxPointRef(line, NxCurvePointRole.Start)));
                    row.Points.Add(points.PointId(new NxPointRef(line, NxCurvePointRole.End)));
                    return row;
                case NxConstraintKind.Parallel:
                case NxConstraintKind.Perpendicular:
                case NxConstraintKind.Collinear:
                case NxConstraintKind.EqualLength:
                case NxConstraintKind.Concentric:
                case NxConstraintKind.EqualRadius:
                case NxConstraintKind.Tangent:
                    foreach (long id in c.Curves)
                    {
                        row.Curves.Add(entityIds[id]);
                    }

                    return row;
                case NxConstraintKind.PointOnLine:
                case NxConstraintKind.Midpoint:
                    row.Points.Add(points.PointId(c.Points[0]));
                    row.Curves.Add(entityIds[c.Curves[0]]);
                    return row;
                case NxConstraintKind.Fix:
                    row.Points.Add(points.PointId(c.Points[0]));
                    return row;
                default:
                    _report.Unsupported("sketch-constraint", c.Kind.ToString());
                    return null;
            }
        }

        private void AddDimensions(
            NxSketch sketch, SketchPointTable points, Dictionary<long, int> entityIds, SketchData data)
        {
            foreach (NxSketchDimension d in sketch.Dimensions)
            {
                data.Dimensions.Add(BuildDimension(d, points, entityIds));
            }
        }

        private static DimensionData BuildDimension(
            NxSketchDimension d, SketchPointTable points, Dictionary<long, int> entityIds)
        {
            var row = new DimensionData
            {
                Kind = DimensionName(d.Kind),
                Expression = d.Expression,
                Driven = d.Driven ? true : (bool?)null,
            };
            if (d.Kind == NxDimensionKind.Distance)
            {
                foreach (NxPointRef p in d.Points)
                {
                    row.Points.Add(points.PointId(p));
                }
            }
            else
            {
                foreach (long id in d.Curves)
                {
                    row.Curves.Add(entityIds[id]);
                }
            }

            return row;
        }

        private static double[] Scale(double[] v) => new[] { v[0] * MmToCm, v[1] * MmToCm, v[2] * MmToCm };

        private static string KindName(NxCurveKind kind)
        {
            switch (kind)
            {
                case NxCurveKind.Line: return "line";
                case NxCurveKind.Circle: return "circle";
                default: return "arc";
            }
        }

        private static string ConstraintName(NxConstraintKind kind)
        {
            switch (kind)
            {
                case NxConstraintKind.Coincident: return "coincident";
                case NxConstraintKind.Horizontal: return "horizontal";
                case NxConstraintKind.Vertical: return "vertical";
                case NxConstraintKind.Parallel: return "parallel";
                case NxConstraintKind.Perpendicular: return "perpendicular";
                case NxConstraintKind.Collinear: return "collinear";
                case NxConstraintKind.EqualLength: return "equalLength";
                case NxConstraintKind.Concentric: return "concentric";
                case NxConstraintKind.EqualRadius: return "equalRadius";
                case NxConstraintKind.Tangent: return "tangent";
                case NxConstraintKind.PointOnLine: return "pointOnLine";
                case NxConstraintKind.Midpoint: return "midpoint";
                case NxConstraintKind.Fix: return "fix";
                default: return kind.ToString();
            }
        }

        private static string DimensionName(NxDimensionKind kind)
        {
            switch (kind)
            {
                case NxDimensionKind.Distance: return "distance";
                case NxDimensionKind.Radius: return "radius";
                case NxDimensionKind.Diameter: return "diameter";
                default: return "angle";
            }
        }
    }
}

File: python/oblikovati_exporter_nx/translate/sketch_translator.py
# SPDX-License-Identifier: GPL-2.0-only
"""Translates one NX sketch into an Oblikovati SketchData.

Shared points (via SketchPointTable), curve entities, geometric constraints, and
parameter-linked dimensions. Lengths convert from the IR's millimetres to the recipe's
centimetre database unit.
"""
from __future__ import annotations

from typing import Dict, List, Optional

from ..model.sketch import (
    NxConstraintKind,
    NxCurve,
    NxCurveKind,
    NxCurvePointRole,
    NxDimensionKind,
    NxPointRef,
    NxSketch,
    NxSketchConstraint,
    NxSketchDimension,
)
from ..recipe.sketch import (
    ConstraintData,
    DimensionData,
    EntityData,
    PlaneData,
    SketchData,
)
from .id_allocator import IdAllocator
from .report import ExportReport
from .sketch_point_table import SketchPointTable
from .units import MM_TO_CM, scale_point

# Constraint kinds whose operands are two-or-more curve ids.
_CURVE_CONSTRAINTS = frozenset(
    {
        NxConstraintKind.PARALLEL,
        NxConstraintKind.PERPENDICULAR,
        NxConstraintKind.COLLINEAR,
        NxConstraintKind.EQUAL_LENGTH,
        NxConstraintKind.CONCENTRIC,
        NxConstraintKind.EQUAL_RADIUS,
        NxConstraintKind.TANGENT,
    }
)


class SketchTranslator:
    def __init__(self, ids: IdAllocator, report: ExportReport) -> None:
        self._ids = ids
        self._report = report

    def translate(self, sketch: NxSketch, sketch_id: int) -> SketchData:
        points = SketchPointTable()
        points.build(sketch, self._ids)

        data = SketchData(
            id=sketch_id,
            name=sketch.name or None,
            plane=_translate_plane(sketch),
            points=list(points.points),
        )
        entity_ids = self._add_entities(sketch, points, data)
        self._add_constraints(sketch, points, entity_ids, data)
        self._add_dimensions(sketch, points, entity_ids, data)
        return data

    def _add_entities(
        self, sketch: NxSketch, points: SketchPointTable, data: SketchData
    ) -> Dict[int, int]:
        entity_ids: Dict[int, int] = {}
        for curve in sketch.curves:
            entity_id = self._ids.next()
            entity_ids[curve.id] = entity_id
            data.entities.append(_build_entity(entity_id, curve, points))
        return entity_ids

    def _add_constraints(
        self,
        sketch: NxSketch,
        points: SketchPointTable,
        entity_ids: Dict[int, int],
        data: SketchData,
    ) -> None:
        for constraint in sketch.constraints:
            row = self._build_constraint(constraint, points, entity_ids)
            if row is not None:
                data.constraints.append(row)

    def _build_constraint(
        self,
        constraint: NxSketchConstraint,
        points: SketchPointTable,
        entity_ids: Dict[int, int],
    ) -> Optional[ConstraintData]:
        kind = constraint.kind
        row = ConstraintData(kind=kind.value)
        if kind == NxConstraintKind.COINCIDENT:
            row.points.append(points.point_id(constraint.points[0]))
            row.points.append(points.point_id(constraint.points[1]))
            return row
        if kind in (NxConstraintKind.HORIZONTAL, NxConstraintKind.VERTICAL):
            # NX applies these to a line; Oblikovati constrains its two endpoints.
            line = constraint.curves[0]
            row.points.append(points.point_id(NxPointRef(line, NxCurvePointRole.START)))
            row.points.append(points.point_id(NxPointRef(line, NxCurvePointRole.END)))
            return row
        if kind in _CURVE_CONSTRAINTS:
            for curve_id in constraint.curves:
                row.curves.append(entity_ids[curve_id])
            return row
        if kind in (NxConstraintKind.POINT_ON_LINE, NxConstraintKind.MIDPOINT):
            row.points.append(points.point_id(constraint.points[0]))
            row.curves.append(entity_ids[constraint.curves[0]])
            return row
        if kind == NxConstraintKind.FIX:
            row.points.append(points.point_id(constraint.points[0]))
            return row
        self._report.unsupported("sketch-constraint", kind.name)
        return None

    def _add_dimensions(
        self,
        sketch: NxSketch,
        points: SketchPointTable,
        entity_ids: Dict[int, int],
        data: SketchData,
    ) -> None:
        for dimension in sketch.dimensions:
            data.dimensions.append(_build_dimension(dimension, points, entity_ids))


def _translate_plane(sketch: NxSketch) -> PlaneData:
    return PlaneData(
        origin=scale_point(sketch.origin),
        xaxis=list(sketch.xaxis),
        yaxis=list(sketch.yaxis),
    )


def _build_entity(entity_id: int, curve: NxCurve, points: SketchPointTable) -> EntityData:
    entity = EntityData(
        id=entity_id,
        kind=curve.kind.value,
        construction=True if curve.construction else None,
    )
    if curve.kind == NxCurveKind.LINE:
        entity.points.append(points.point_id(NxPointRef(curve.id, NxCurvePointRole.START)))
        entity.points.append(points.point_id(NxPointRef(curve.id, NxCurvePointRole.END)))
        entity.centerline = True if curve.centerline else None
    elif curve.kind == NxCurveKind.CIRCLE:
        entity.points.append(points.point_id(NxPointRef(curve.id, NxCurvePointRole.CENTER)))
        entity.radius = curve.radius * MM_TO_CM
    else:  # arc
        entity.points.append(points.point_id(NxPointRef(curve.id, NxCurvePointRole.CENTER)))
        entity.points.append(points.point_id(NxPointRef(curve.id, NxCurvePointRole.START)))
        entity.points.append(points.point_id(NxPointRef(curve.id, NxCurvePointRole.END)))
        entity.ccw = True if curve.ccw else None
    return entity


def _build_dimension(
    dimension: NxSketchDimension,
    points: SketchPointTable,
    entity_ids: Dict[int, int],
) -> DimensionData:
    row = DimensionData(
        kind=dimension.kind.value,
        expression=dimension.expression,
        driven=True if dimension.driven else None,
    )
    if dimension.kind == NxDimensionKind.DISTANCE:
        for point_ref in dimension.points:
            row.points.append(points.point_id(point_ref))
    else:
        for curve_id in dimension.curves:
            row.curves.append(entity_ids[curve_id])
    return row
